pkg/controllers/user/pipeline/controller/pipelineexecution/statesyncer.go

```diff
@@ -57,56 +54,80 @@ func (s *ExecutionStateSyncer) syncState() {
+
+	for _, originE := range executions {
+		e := originE.DeepCopy()
+		if v3.PipelineExecutionConditionInitialized.IsUnknown(e) {
+			ready, err := s.pipelineEngine.PreCheck(e)
+			if err != nil {
+				v3.PipelineExecutionConditionBuilt.False(e)
+				v3.PipelineExecutionConditionBuilt.ReasonAndMessageFromError(e, err)
+				e.Status.ExecutionState = utils.StateFailed
+			}
+			v3.PipelineExecutionConditionInitialized.Message(e, "setting up jenkins")
+			if ready {
+
+				if err := s.pipelineEngine.RunPipelineExecution(e); err != nil {
+					v3.PipelineExecutionConditionProvisioned.False(e)
+					v3.PipelineExecutionConditionProvisioned.ReasonAndMessageFromError(e, err)
+					e.Status.ExecutionState = utils.StateFailed
+					if err := s.updateExecutionAndLastRunState(e); err != nil {
+						logrus.Error(err)
+					}
+					continue
+				}
+				v3.PipelineExecutionConditionInitialized.True(e)
+				v3.PipelineExecutionConditionProvisioned.CreateUnknownIfNotExists(e)
+				v3.PipelineExecutionConditionProvisioned.Message(e, "Assigning jobs to pipeline engine")
+			}
+
 			if err := s.updateExecutionAndLastRunState(e); err != nil {
-				logrus.Errorf("Error update pipeline execution - %v", err)
-				return
+				logrus.Error(err)
+				continue
 			}
-		}
-		return
-	}
-	for _, e := range executions {
-		if e.Status.ExecutionState == utils.StateWaiting || e.Status.ExecutionState == utils.StateBuilding {
+		} else if v3.PipelineExecutionConditionInitialized.IsTrue(e) {
 			updated, err := s.pipelineEngine.SyncExecution(e)
 			if err != nil {
-				logrus.Errorf("Error sync pipeline execution - %v", err)
-				e.Status.ExecutionState = utils.StateFail
-				v3.PipelineExecutionConditonProvisioned.False(e)
-				v3.PipelineExecutionConditonProvisioned.ReasonAndMessageFromError(e, err)
+				logrus.Errorf("got error in syncExecution: %v", err)
+				v3.PipelineExecutionConditionBuilt.False(e)
+				v3.PipelineExecutionConditionBuilt.ReasonAndMessageFromError(e, err)
+				e.Status.ExecutionState = utils.StateFailed
 				updated = true
 			}
 			if updated {
 				if err := s.updateExecutionAndLastRunState(e); err != nil {
-					logrus.Errorf("Error update pipeline execution - %v", err)
-					return
+					logrus.Error(err)
+					continue
 				}
 			}
 		} else {
 			if err := s.updateExecutionAndLastRunState(e); err != nil {
 				logrus.Errorf("Error update pipeline execution - %v", err)
-				return
 			}
 		}
 	}
+
 	logrus.Debugf("Sync pipeline execution state complete")
 }
 
 func (s *ExecutionStateSyncer) updateExecutionAndLastRunState(execution *v3.PipelineExecution) error {
-	if execution.Status.ExecutionState != utils.StateWaiting && execution.Status.ExecutionState != utils.StateBuilding {
+
+	if v3.PipelineExecutionConditionInitialized.IsFalse(execution) || v3.PipelineExecutionConditionProvisioned.IsFalse(execution) ||
+		v3.PipelineExecutionConditionBuilt.IsFalse(execution) {
 		execution.Labels[utils.PipelineFinishLabel] = "true"
+
```

Comment:
why all this conditions are false and we mark the label finish is true?

Reply:
When any of the condition is false we consider the execution fails for some reason and mark it as finished.